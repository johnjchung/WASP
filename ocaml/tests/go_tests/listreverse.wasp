Func list<int> reverseList(list<int> a) {
    int i = sizeof a - 1;
    list<int> reverseList = [0] @ 0->0;
    while (i >= 0) {
        reverseList = reverseList ++ a @ i->(i+1);
        i = i - 1;
    }
    return reverseList;
}

Endpoint reverse(GET:) {
    list<int> sublist = reverseList([1,2,3,4,5,6,7,8,9,10]);
}
