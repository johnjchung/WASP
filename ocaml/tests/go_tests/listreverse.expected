package main

import (
	"github.com/gin-gonic/gin"
	"math"
	"os"
)

func int_conv(anything interface{}) int {
	switch v := anything.(type) {
		case int:
			return v
		case float64, float32:
			var float_var = v.(float64)
			return int(math.Trunc(float_var))
		default:
			return 0
	}
	return 0
}

func float_conv(anything interface{}) float64 {
 	 switch v := anything.(type) {
		case int:
            return float64(v)
		case float64:
			return v
		default:
			return 0
	}
	return 0.0
}

type reverse_struct struct {
	Sublist []int `db:"sublist" json:"sublist"`
}

func main() {
	router := gin.Default()
	router.GET("/reverse", Get_reverse)
	router.Run(":"+os.Getenv("PORT"))
}

func reverseList(a []int) []int {
	i := int_conv(len(a)) - int_conv(1)
	reverseList := []int {0}[0:0]
	for i >= 0 {
	reverseList = append(reverseList, a[i:int_conv(i) + int_conv(1)]...)
	i = int_conv(i) - int_conv(1)
	}
	return reverseList
}

func reverse_struct_wrapper() reverse_struct {
	return reverse_struct{Sublist: reverseList([]int {1, 2, 3, 4, 5, 6, 7, 8, 9, 10})}
}

func Get_reverse(c *gin.Context) {
	reverse_struct := reverse_struct_wrapper()
	c.JSON(200, reverse_struct)
}
